### PR TITLE
libnetwork: improve embedded DNS resolver

### LINKED ISSUE #44575: DNS queries fails for Amazon S3 domains since Nov 19
### Description

On November 19 DNS queries for Amazon S3 from inside our containers started failing. It seems to be related to short TTL and multiple IP addresses being returned for A records for Amazon S3 domains.

We get this error when connections are being opened inside containers: `getaddrinfo: No address associated with hostname`.

Amazon has TTL of 5 seconds set for their S3 domains. On November 19 they changed their DNS records to return multiple IP addresses for single domain. You can see this change on screenshot from [DNS history here](https://gist.githubusercontent.com/vojtad/3bac63b8c91b1ec50e8d8b36047317fa/raw/b9e81d1ec20e2e8e32ad4b418869a26d3bfdaf90/screencapture-securitytrails-domain-s3-amazonaws-com-history-a-2022-12-03-18_40_08.png).

Our DNS containers use Docker resolver on 127.0.0.11. Host system uses caching DNS server on our Mikrotik router. Changing host system to use another layer of caching using dnsmasq or systemd-resolved didn't help. It even made the error happen more frequently.

### Reproduce

1. Set up dnsmasq on host system (optional but makes the error more consistent).
1. Create custom network to make Docker use its resolver using `docker network create --driver bridge dnserror-net`.
2. Start the container using `docker run -it --network dnserror-net ubuntu:jammy bash -l`.
3. Build a tool for quering Amazon S3 domain in parallel (makes the error happend more consistently).
  1. Install build-essentials using `apt-get update && apt-get install -y build-essential`.
  2. Save [source code](https://gist.github.com/vojtad/a3b4f49f0a5d16bcda9b0ba5e5600a2a) in `main.c`.
  3. Compile the tool using `gcc main.c`.
4. Run the tool using `./a.out 1`. The argument is number of threads to use.
5. After some time (usually couple of seconds) queries start to fail. They recover again after couple of seconds.

### Expected behavior

I expect DNS queries to work regardless what DNS server host system uses.

### docker version

```bash
Client: Docker Engine - Community
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 18:02:28 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       3056208
  Built:            Tue Oct 25 18:00:19 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.10
  GitCommit:        770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  scan: Docker Scan (Docker Inc., v0.17.0)

Server:
 Containers: 8
  Running: 5
  Paused: 0
  Stopped: 3
 Images: 4
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2 io.containerd.runtime.v1.linux nvidia
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 770bd0108c32f3fb5c73ae1264f7e503fe7b2661
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: default
  cgroupns
 Kernel Version: 5.16.0-0.bpo.4-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 24
 Total Memory: 62.78GiB
 Name: ewok-001
 ID: HZRK:QO5C:33RJ:LFTC:MNM5:24CL:PC7M:IPKR:BYBA:J7SO:2YSH:GLKQ
 Docker Root Dir: /fast/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

Environment:
1. Host system is updated Debian stable.
2. Docker container is `ubuntu:jammy`
3. Docker version is `Docker version 20.10.21, build baeda1f`

tcpdumps (you can open them in Wireshark)
- [tcpdump from host ethernet connection](https://gist.github.com/vojtad/3bac63b8c91b1ec50e8d8b36047317fa/raw/7d75eb3d3448381bf252ae55ea5123a132c46658/host.pcap)
- [tcpdump from Docker container](https://gist.github.com/vojtad/3bac63b8c91b1ec50e8d8b36047317fa/raw/7d75eb3d3448381bf252ae55ea5123a132c46658/docker.pcap)

From tcpdump from Docker container you can see Docker resolver replies with empty DNS response even though tcpdump from host connections doesn't show any empty response at all.

When I change dnsmasq or systemd-resolved to use `1.1.1.1` instead of our Mikrotik router it seems to fix the issue. However, I need to use my router as DNS server. I don't see any invalid DNS or any other issues in Wireshark when DNS is set to Mikrotik router. I guess the issue is some kind of a race condition on Docker resolver when TTL expires and router responds with records with both TTL set to 0 and to 5 seconds.

### LINKED ISSUE #44610: Improve logs for DNS failures
```
level=debug msg="Name To resolve: s3.amazonaws.com."
level=debug msg="[resolver] query s3.amazonaws.com. (A) from 172.20.0.2:55963, forwarding to udp:10.155.0.1"
level=debug msg="[resolver] failed to read from DNS server" error="dns: overflowing header size"
```

We should probably consider updating some of those logs to not be "debug" to make them more discoverable, and for that last one

- include the DNS server where the overflowing header size came from.
- include the query as well?

_Originally posted by @thaJeztah in https://github.com/moby/moby/issues/44575#issuecomment-1342395724_

### LINKED ISSUE #44979: Invalid PTR query causes resolver panic
### Description

`dockerd` crashed with this trace: 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x55a9fa3227df]

goroutine 306 [running]:
github.com/docker/docker/libnetwork.(*resolver).ServeDNS(0xc0009e7b20, {0x55a9fb8b90e0, 0xc000a0f300}, 0xc0012297a0)
        /go/src/github.com/docker/docker/libnetwork/resolver.go:399 +0x6ff
github.com/docker/docker/vendor/github.com/miekg/dns.(*Server).serveDNS(0xc000eea6c0, {0xc000b8a800, 0x31, 0x200}, 0xc000a0f300)
        /go/src/github.com/docker/docker/vendor/github.com/miekg/dns/server.go:651 +0x4e2
github.com/docker/docker/vendor/github.com/miekg/dns.(*Server).serveUDPPacket(0xc000eea6c0, 0xc000eea7e0?, {0xc000b8a800, 0x31, 0x200}, {0x55a9fb8b5af0?, 0xc000d6c9d8}, 0xc001232f60, {0x0, 0x0})
        /go/src/github.com/docker/docker/vendor/github.com/miekg/dns/server.go:591 +0x185
created by github.com/docker/docker/vendor/github.com/miekg/dns.(*Server).serveUDP
        /go/src/github.com/docker/docker/vendor/github.com/miekg/dns/server.go:521 +0x485
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x55a9fa3227df]

goroutine 271 [running]:
github.com/docker/docker/libnetwork.(*resolver).ServeDNS(0xc0009e7b20, {0x55a9fb8b90e0, 0xc0011bb400}, 0xc000fb2ab0)
        /go/src/github.com/docker/docker/libnetwork/resolver.go:399 +0x6ff
github.com/docker/docker/vendor/github.com/miekg/dns.(*Server).serveDNS(0xc000eea6c0, {0xc0006d2000, 0x35, 0x200}, 0xc0011bb400)
        /go/src/github.com/docker/docker/vendor/github.com/miekg/dns/server.go:651 +0x4e2
github.com/docker/docker/vendor/github.com/miekg/dns.(*Server).serveUDPPacket(0xc000eea6c0, 0x55a9fb88bbc0?, {0xc0006d2000, 0x35, 0x200}, {0x55a9fb8b5af0?, 0xc000d6c9d8}, 0xc001232f00, {0x0, 0x0})
        /go/src/github.com/docker/docker/vendor/github.com/miekg/dns/server.go:591 +0x185
created by github.com/docker/docker/vendor/github.com/miekg/dns.(*Server).serveUDP
        /go/src/github.com/docker/docker/vendor/github.com/miekg/dns/server.go:521 +0x485
```

### Reproduce

I'm not totally sure where these bad DNS queries originate from yet :sweat_smile:  but these PTR queries very consistently crash the docker daemon. They look something like this:
 ```
invalid PTR query, lb._dns-sd._udp.foo.er0k.net.
invalid PTR query, lb._dns-sd._udp.bar.er0k.net.
```

It looks like when the resolver encounters an error in `ServeDNS()`, it [tries to log the error](
https://github.com/moby/moby/blob/master/libnetwork/resolver.go#L397) along with the local address of the external connection, but `extConn` is `nil` 

### Expected behavior

dockerd should log the error and not crash

### docker version

```bash
Client: Docker Engine - Community
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:46:54 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       bc3805a
  Built:            Thu Feb  9 19:46:54 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.16
  GitCommit:        31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.10.0-20-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 473.2MiB
 Name: bug
 ID: 12a513eb-f4d8-4ebf-a995-75d7ade340b5
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 22
  Goroutines: 32
  System Time: 2023-02-11T19:22:25.521287084Z
  EventsListeners: 0
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

### Additional Info

_No response_

### PR DESCRIPTION
**Pull request by corhere**
- Fixes #44575 
- Fixes #44610 
- Fixes #44979

**- What I did**
Lots. See the individual commits for details.

**- How I did it**
Read The Fine RFCs.

**- How to verify it**
New unit tests.

**- Description for the changelog**

- Many improvements to the quality of the embedded DNS resolver used in non-default container networks

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by corhere** [comment]
Note to self: address https://github.com/moby/moby/pull/44491#discussion_r1060111679 and https://github.com/moby/moby/pull/44491#discussion_r1060113301

File: libnetwork/resolver.go [hunk 1]
```diff
@@ -393,199 +363,200 @@ func (r *resolver) ServeDNS(w dns.ResponseWriter, query *dns.Msg) {
...
-		for i := 0; i < maxExtDNS; i++ {
-			extDNS := &r.extDNSList[i]
-			if extDNS.IPStr == "" {
-				break
-			}
-			extConnect := func() {
-				addr := fmt.Sprintf("%s:%d", extDNS.IPStr, 53)
-				extConn, err = net.DialTimeout(proto, addr, extIOTimeout)
-			}
+		addr := fmt.Sprintf("%s:%d", server.IPStr, server.port)
```
Comment (corhere) [inline libnetwork/resolver.go:427]: This'll do the wrong thing with IPv6 addresses. Use `net.JoinHostPort`

File: libnetwork/resolver.go [hunk 2]
```diff
@@ -543,7 +543,19 @@ func (r *resolver) exchange(proto string, extDNS extDNSEntry, query *dns.Msg) *d
 	})
 	log.Debug("[resolver] forwarding query")
 
-	resp, _, err := (&dns.Client{Timeout: extIOTimeout}).ExchangeWithConn(query, &dns.Conn{Conn: extConn})
+	resp, _, err := (&dns.Client{
```
Comment (cpuguy83) [inline libnetwork/resolver.go:532]: Maybe this commit can be merged with where this client was introduced?

**Review by cpuguy83 — APPROVED** [review]

File: libnetwork/resolver.go (hunk 1 above)
Comment (corhere) [reply libnetwork/resolver.go:427]: On second thought, I would rather leave this change for another PR dedicated to adding support for external IPv6 DNS servers. Every code path which sets the external servers filters out IPv6 addresses.
https://github.com/moby/moby/blob/4e276400760585c9d5969488c9b44e5176a58c18/libnetwork/sandbox_dns_unix.go#L261
https://github.com/moby/moby/blob/4e276400760585c9d5969488c9b44e5176a58c18/libnetwork/sandbox_dns_unix.go#L267
https://github.com/moby/moby/blob/4e276400760585c9d5969488c9b44e5176a58c18/libnetwork/sandbox_dns_unix.go#L402

**Review by neersighted — APPROVED** [review]

**Pull request merged**